### DISCUSSION TITLE
Where's the CASCADIA_HOSTING_CLASS_WINDOW window?

### DISCUSSION THREAD

**Discussion by vefatica** [microsoft/terminal#14492] [Ideas]
I replied to  "DISCUSSION: Is my app running in Terminal? #7434" but that reply isn't showing up anywhere (excuse me if I'm impatient).

With specific regard to a shell wanting to know where the Windows Terminal user interface is ...

I revisit this every now and then.  I'm using the process ID of GetConsoleWindow() and a Toolhelp32Snapshot to find the parent process of OpenConsole.exe.  Then I use EnumWindows() to find a CASCADIA_HOSTING_WINDOW_CLASS window belonging to that parent.  It works, at least to my satisfaction, but it's a bit cumbersome.

Could you make the handle of the CASCADIA_HOSTING_WINDOW_CLASS window more readily available?  One thought is to put it in the value of the WT_SESSION environment variable (along with the GUID which is (AFAICT) useless to the shell.

`WT_SESSION=69f813f0-06ef-49d5-a64a-95359e0305c8, 0x303C0`

A shell might be looking for the existence of that variable anyway.  If it's value had more useful information in it, that would be great.

Am I wrong about the GUID?  Is it of any value to the shell (or to anyone)?

**Comment by zadjii-msft** [comment]
Alrighty, there's two bits here:

* "I want to get at the Terminal's HWND"
  * I'm EXTREMELY reluctant to make this easier for anyone. Messing with the Console's HWND has only ever created compatibility issues that have been extremely challenging for us to deal with. I'd rather not repeat the mistakes of the past in that regard. 
* Is `WT_SESSION` used for anything?
  * People are using it for.. various things. IIRC some people use it in scripts to change the current Terminal's settings? That feels like a mild fact. [It's also not reliable](https://github.com/microsoft/terminal/issues/13006), for a number of reasons.

**Comment by vefatica** [comment]
I don't want to _mess_ with it; I only want to use it to position GUI thingies nicely with respect to the user interface. IIRC, others wanted to do the same thing.

You can get that HWND anyway.  It's just the window of a user interface, hardly the sort of thing Windows keeps a secret. I don't see the big deal in making getting it easy.

**Comment by MitchBodmer** [comment]
@zadjii-msft one reason we might want to get the window handle is to [hand it over to the WAM broker](https://learn.microsoft.com/en-us/entra/identity-platform/scenario-desktop-acquire-token-wam#console-applications). In my neck of the woods this is the only approved way to authenticate console apps. Their official documentation linked above recommends the "not recommended" solution of using `GetConsoleWindow` with `GetAncestor`.

**Reply by zadjii-msft** [reply]
Where does it ever say that that is not the recommended way to do it? That's literally the way I recommended the WAM team to do that 🤣

**Reply by zadjii-msft** [reply]
For notes on my quest, see https://github.com/microsoft/terminal/issues/2988#issuecomment-1044501169

**Reply by vefatica** [reply]
That was also recommended in this forum, quite possibly by @zadjii-msft.  I have used it for several years.

```
HWND GetCascadiaWindow()
{
	WCHAR szClass[32];
	HWND hWnd = GetAncestor(GetConsoleWindow(), GA_ROOTOWNER);

	if	(
			hWnd == nullptr
			|| GetClassNameW(hWnd, szClass, 32) == 0
			|| lstrcmpW(szClass, L"CASCADIA_HOSTING_WINDOW_CLASS") != 0
		)
		return nullptr;

	return hWnd;
}
```

**Reply by MitchBodmer** [reply]
@zadjii-msft by "not recommended" I'm referring to the header on the documentation for `GetConsoleWindow`.
<img width="680" height="421" alt="image" src="https://github.com/user-attachments/assets/06974bb0-9ec0-45eb-8575-3c2e8b3f8b79" />
That's what I was trying to imply with the scare quotes. If this is the only recommended way to get the handle, then maybe we should consider removing that header in the `GetConsoleWindow` docs.

**Comment by vefatica** [comment]
Hmmm!  I may have spoke too soon.  Why does this function return nullptr when I'm in the debug tap?

```
HWND GetCascadiaWindow()
{
	WCHAR szClass[32];
	HWND hWnd = GetAncestor(GetConsoleWindow(), GA_ROOTOWNER);

	if	(
			hWnd == nullptr
			|| GetClassNameW(hWnd, szClass, 32) == 0
			|| lstrcmpW(szClass, L"CASCADIA_HOSTING_WINDOW_CLASS") != 0
		)
		return nullptr;

	return hWnd;
}
```

**Reply by DHowett** [reply]
Do not use the debug tap for production purposes. I'll be forced to remove it.

**Comment by vefatica** [comment]
What good would it be for production?  I'm using it for debugging.

**Comment by vefatica** [comment]
Can you say why my function returns nullptr when TCC is in the debug tap?

**Reply by DHowett** [reply]
Not outside of business hours, I can't!

**Reply by vefatica** [reply]
Never mind; I figured it out.
